core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java

```diff
@@ -0,0 +1,241 @@
+              writer.startList(SqlWriter.FrameTypeEnum.FUN_CALL, "(", ")");
+          for (SqlNode operand : call.getOperandList()) {
+            writer.sep(",");
+            operand.unparse(writer, 0, 0);
+          }
+          writer.endList(frame);
+        }
+      };
+
+  /** Creates a ClickHouseSqlDialect. */
+  public ClickHouseSqlDialect(Context context) {
+    super(context);
+  }
+
+  @Override public boolean supportsCharSet() {
+    return false;
+  }
+
+  @Override public boolean supportsNestedAggregations() {
+    return false;
+  }
+
+  @Override public boolean supportsWindowFunctions() {
+    return false;
+  }
+
+  @Override public CalendarPolicy getCalendarPolicy() {
+    return CalendarPolicy.SHIFT;
+  }
+
+  @Override public SqlNode getCastSpec(RelDataType type) {
+    String castSpec;
+    switch (type.getSqlTypeName()) {
+    case VARCHAR:
+      castSpec = "String";
+      break;
+    case TINYINT:
+      castSpec = "Int8";
+      break;
+    case SMALLINT:
+      castSpec = "Int16";
+      break;
+    case INTEGER:
+      castSpec = "Int32";
+      break;
+    case BIGINT:
+      castSpec = "Int64";
+      break;
+    case FLOAT:
+      castSpec = "Float32";
+      break;
+    case DOUBLE:
+      castSpec = "Float64";
+      break;
+    case DATE:
+      castSpec = "Date";
+      break;
+    case TIMESTAMP:
+    case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      castSpec = "DateTime";
+      break;
+    default:
+      return super.getCastSpec(type);
+    }
+
+    return new SqlDataTypeSpec(new SqlIdentifier(castSpec, SqlParserPos.ZERO),
+        type.getPrecision(), -1, null, null, SqlParserPos.ZERO);
+  }
+
+  @Override public void unparseDateTimeLiteral(SqlWriter writer,
+      SqlAbstractDateTimeLiteral literal, int leftPrec, int rightPrec) {
+    String toFunc;
+    if (literal instanceof SqlDateLiteral) {
+      toFunc = "toDate";
+    } else if (literal instanceof SqlTimestampLiteral) {
+      toFunc = "toDateTime";
+    } else if (literal instanceof SqlTimeLiteral) {
+      toFunc = "toTime";
+    } else {
+      throw new RuntimeException("ClickHouse does not support DateTime literal: "
+          + literal);
+    }
+
+    writer.literal(toFunc + "('" + literal.toFormattedString() + "')");
+  }
+
+  @Override public void unparseOffsetFetch(SqlWriter writer, SqlNode offset,
+      SqlNode fetch) {
+    Preconditions.checkArgument(fetch != null);
+
+    writer.newlineAndIndent();
+    final SqlWriter.Frame frame =
+        writer.startList(SqlWriter.FrameTypeEnum.FETCH);
+    writer.keyword("LIMIT");
+
+    if (offset != null) {
+      offset.unparse(writer, -1, -1);
+      writer.sep(",", true);
+    }
+
+    fetch.unparse(writer, -1, -1);
+    writer.endList(frame);
+  }
+
+  @Override public void unparseCall(SqlWriter writer, SqlCall call,
+      int leftPrec, int rightPrec) {
+    if (call.getOperator() == SqlStdOperatorTable.SUBSTRING) {
+      CLICKHOUSE_SUBSTRING.unparse(writer, call, 0, 0);
+    } else {
```

Comment:
Why we must have a `CLICKHOUSE_SUBSTRING ` operator there ?

Reply:
The clickhouse substring function looks like this: `substring(s, offset, length)` which is a bit different from the calcite `SUBSTRING(string FROM integer FOR integer)`

Reply:
I mean just to unparse with CLICKHOUSE style when we encounter the `SqlStdOperatorTable.SUBSTRING`, seems not necessary with an operator there.

Reply:
It's just a way to keep that logic in one place. But I'm not opposed to calling the writer directly here if you think it's better.